File: leetcode/leetcode746.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost)
    {
        vector<int> dp = cost;
        for (int i = 2; i < dp.size(); i++) {
            dp[i] += min(dp[i - 1], dp[i - 2]);
        }
        return min(dp[dp.size() - 1], dp[dp.size() - 2]);
    }
};
int main(void)
{
    return 0;
}

File: leetcode/100-999/leetcode746.py
from typing import List
from functools import lru_cache


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        @lru_cache
        def chaewon(top: int) -> int:
            if top <= 1:
                return 0
            return min(
                chaewon(top - 1) + cost[top - 1], chaewon(top - 2) + cost[top - 2]
            )

        return chaewon(len(cost))
